apps/els_lsp/src/els_dt_document.erl

```diff
@@ -223,25 +231,42 @@ wrapping_functions(Document, Range) ->
   #{start := #{character := Character, line := Line}} = Range,
   wrapping_functions(Document, Line, Character).
 
--spec find_candidates(binary()) -> [uri()].
+-spec find_candidates(atom() | string()) -> [uri()].
 find_candidates(Pattern) ->
-  %% ets:fun2ms(fun(#els_dt_document{source = Source, uri = Uri, text = Text})
-  %% when Source =/= otp -> {Uri, Text} end).
+  %% ets:fun2ms(fun(#els_dt_document{source = Source, uri = Uri, words = Words})
+  %% when Source =/= otp -> {Uri, Words} end).
   MS = [{#els_dt_document{ uri = '$1'
                          , source = '$2'
                          , buffer = '_'
                          , kind = '_'
-                         , text = '$3'
+                         , text = '_'
                          , md5 = '_'
                          , pois = '_'
+                         , words = '$3'
                          }
         , [{'=/=', '$2', otp}]
         , [{{'$1', '$3'}}]}],
   All = ets:select(name(), MS),
-  Fun = fun({Uri, Text}) ->
-            case binary:matches(Text, Pattern) of
-              [] -> false;
-              _ -> {true, Uri}
+  Fun = fun({Uri, Words}) ->
+            case sets:is_element(Pattern, Words) of
+              true -> {true, Uri};
+              false -> false
             end
         end,
   lists:filtermap(Fun, All).
+
+-spec get_words(binary()) -> sets:set().
+get_words(Text) ->
+  case erl_scan:string(els_utils:to_list(Text)) of
+    {ok, Tokens, _EndLocation} ->
+      Fun = fun({atom, _Location, Atom}, Words) ->
+                sets:add_element(Atom, Words);
+               ({string, _Location, String}, Words) ->
+                sets:add_element(String, Words);
+               (_, Words) ->
+                Words
+            end,
+      lists:foldl(Fun, sets:new(), Tokens);
```

Comment:
`sets` in general is pretty slow and uses a lot of memory, unless you use `sets:new([{version, 2}])` for the map-based representation.

That said, in general it seems like shipping this data back and forth from ets might be really expensive.

An alternative would be to use a plain map to store the data: `#{atom() | string() => true}`

And query it with a match spec like:
```erl
  MS = [{#els_dt_document{ uri = '$1'
                         , source = '$2'
                         , buffer = '_'
                         , kind = '_'
                         , text = '_'
                         , md5 = '_'
                         , pois = '_'
                         , words = '$3'
                         }
        , [{'=/=', '$2', otp}, {'is_map_key', Search, '$3'}]
        , ['$1']}],
```

This way there's no need to copy the data out of ETS

Reply:
Version 2 is only supported from OTP 24, right? Ideally Erlang LS shouldn't care about the project version used by a user, but for the time being we have some assumptions in the code base that force us to be backward compatible. So, will need to add it as an ifdef.

Reply:
Thanks for this! I will leave this as a follow up, so that we can actually measure the eventual improvement.